hubstorage/resourcetype.py

```diff
@@ -65,10 +85,37 @@ def _add_resume_param(self, lastline, offset, params):
     def iter_values(self, *args, **kwargs):
         """Reliably iterate through all data as python objects
 
-        calls iter_json, decoding the results
+        calls either iter_json or iter_msgpack, decoding the results
         """
+        if self._supports_msgpack:
+            return mpdecode(self.iter_msgpack(*args, **kwargs))
         return jldecode(self.iter_json(*args, **kwargs))
 
+    def iter_msgpack(self, _path=None, requests_params=None, **apiparams):
+        """Reliably iterate through all data as json strings"""
+        requests_params = dict(requests_params or {})
+        requests_params.setdefault('method', 'GET')
+        requests_params = self._enforce_msgpack(**requests_params)
+        lastexc = None
+        for attempt in xrange(self.MAX_RETRIES):
+            try:
+                return self._content(_path=_path, params=apiparams,
+                                     **requests_params)
+            except (ValueError, socket.error, rexc.RequestException) as exc:
+                # catch requests exceptions other than HTTPError
+                if isinstance(exc, rexc.HTTPError):
+                    raise
+                lastexc = exc
+                url = urlpathjoin(self.url, _path)
+                msg = 'Retrying read of %s in %ds: attempt=%d/%d error=%s'
+                args = url, self.RETRY_INTERVAL, attempt, self.MAX_RETRIES, exc
+                logger.debug(msg, *args)
+                time.sleep(self.RETRY_INTERVAL)
+        url = urlpathjoin(self.url, _path)
+        logger.error('Failed %d times reading items from %s, params %s, '
+                     'last error was: %s', self.MAX_RETRIES, url,
+                     apiparams, lastexc)
```

Comment:
I assumed BytesIO would stream the data while reading it, but I noticed `requests` downloads everything [1] anyway. I re-factored out the common code. Thanks for pointing it out.

Is there any reason the retry logic isn't done using `retrying`?

[1] https://github.com/kennethreitz/requests/blob/master/requests/models.py#L737